### PR TITLE
Support custom Flyway credentials/URL

### LINKED ISSUE #24639: configure dedicated db user for database migrations: DML-only user for datasource, but DDL user for migration
### Description

My usecase is simple in theory:
- Database migrations (using either flyway or liquibase) are run using a database user with DDL grants
- The datasource itself uses a database user _without_ DDL rights afterwards, so preferably only DML.

This is a safety measurement against theoretical situations where the application goes haywire for any reason. This includes e.g. accidentally setting `quarkus.hibernate-orm.database.generation` to `drop-and-create`. If the application runs with a restricted database user, damage in such situations can be mitigated.

This implies the possibility to configure a separate username-password-pair that is used for the database migration only. Right now this does not seem to be possible out-of-the-box. Two workarounds right now are:

- Configuring an entirely new datasource with different credentials and only setting `migrate-at-start` for that one. This requires a bit of config duplication and fiddling to get right but can be made to work.
- Not using `migrate-at-start` and instead manually migrating in-code: For example when using flyway, if you manually create a `Flyway` object in code you are able to provide a different username and password for the flyway config. But this then doesn't let you use hibernate schema validation anymore, because the flyway migration happens after the validation.

### Implementation ideas

I am assuming the use of flyway again, since that's what I am currently using. But some ideas may also apply to liquibase.

From a high-level view I could imagine there to be two additional configurations:
```
# Name of a custom database user for flyway to use. Must have DDL rights
quarkus.flyway.username=my_schema_owner
# Password of the custom database user for flyway
quarkus.flyway.password=my_secret_password
```

These configs could then be passed through to the `FlywayCreator`, which replaces `configure.dataSource(dataSource);` with `configure.dataSource(jdbcUrl, username, password);` if the custom username and password are set.  
Alternatively, the username and password can be forwarded to the flyway config and flyway itself could be modified to use the `DataSource#getConnection(String username, String password)` overload instead of `DataSource#getConnection()` if it sees a username and password in addition to a datasource in its configuration.

What do you think? Is this a sensible usecase to consider?

### PR DESCRIPTION
**Pull request by mfpc**
- Fixes #24639

### CODE REVIEW

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java
Comment (sberyozkin) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:47]:
```suggestion
     * Sets a specific user that Flyway will use to connect to the datasource.
```

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java
Comment (mfpc) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java:47]: Sure, please give me some time to work on it.

Comment (mfpc) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java:47]: @sberyozkin  I'm get this error below when I try to use CrednetialsProviderFinder.find

> 
> [ERROR] io.quarkus.flyway.runtime.FlywayCreatorTest.testIgnoreMissingMigrations  Time elapsed: 0.022 s  <<< ERROR!
> java.lang.NullPointerException: Cannot invoke "io.quarkus.arc.ArcContainer.instance(java.lang.Class, java.lang.annotation.Annotation[])" because "container" is null
> 	at io.quarkus.credentials.runtime.CredentialsProviderFinder.find(CredentialsProviderFinder.java:13)
> 	at io.quarkus.flyway.runtime.FlywayCreator.createFlyway(FlywayCreator.java:48)
> 	at io.quarkus.flyway.runtime.FlywayCreatorTest.createdFlywayConfig(FlywayCreatorTest.java:285)
> 	at io.quarkus.flyway.runtime.FlywayCreatorTest.testIgnoreMissingMigrations(FlywayCreatorTest.java

:206

Could you give me a explanation how I should implement this?

Thank you !

Comment (mfpc) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java:47]: Hi,
I added the code below, but i'm facing problem above with ArcContainer ? I think there is more steps to get the container ready, could you give some help explain that? @sberyozkin 

>    String beanName = flywayRuntimeConfig.credentialsProviderName.orElse(null);
>             CredentialsProvider credentialsProvider = CredentialsProviderFinder.find(beanName);
>             Map<String, String> mapCredentials = credentialsProvider.getCredentials(null);
>             if (mapCredentials.get(CredentialsProvider.USER_PROPERTY_NAME) != null &&
>                     mapCredentials.get(CredentialsProvider.PASSWORD_PROPERTY_NAME) != null) {
>                 configure.dataSource(
>                         new CredentialsDataSource(dataSource, mapCredentials.get(CredentialsProvider.USER_PROPERTY_NAME),
>                                 mapCredentials.get(CredentialsProvider.PASSWORD_PROPERTY_NAME)));

Comment (gastaldi) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java:47]: I think this can be implemented in a separate PR once there is a demand for this use case. For now, I'd recommend keeping it simple for now.

**Comment by gastaldi** [comment]
The problem with closing PRs for the same issue is that you miss important comments. I think this is what we're looking for: https://github.com/quarkusio/quarkus/pull/29628#discussion_r1038226937

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java
Comment (sberyozkin) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java:47]: Sorry for a delay, @mfpc and thanks for giving it a try :+1:  
Thanks @gastaldi, makes sense

**Review by sberyozkin — APPROVED** [review]

**Review by gsmet — COMMENTED** [review]
Just a note that I want to have a look at this before we merge.

**Review by gastaldi — CHANGES_REQUESTED** [review]
There are some stuff missing. I'll meet with @mfpc later this week to finish it

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/CredentialsDataSource.java
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/CredentialsDataSource.java:14]: This class is no longer used and may be deleted

File: extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionCredentialsDatasourceTest.java
Comment (gastaldi) [inline extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionCredentialsDatasourceTest.java:35]: I personally don't think this test has much value. I'd be happy if the test could do something like:
- A Flyway script that inserts the current user in a specific table (eg. using the `USER()` function)
- The JAX-RS resource that has access to the Database executes a query returning the data in the previous table
- The test invokes the JAX-RS resource and compares if the result from the previous call matches what's returned by `fixture.username("")`

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:76]: Unless you're implementing what @sberyozkin proposed in https://github.com/quarkusio/quarkus/pull/29756/files#r1043347239 I'd suggest to remove these lines

**Review by gastaldi — CHANGES_REQUESTED** [review]

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java:55]: The `unwrap` method in `io.agroal.pool.DataSource` always returns null. Better cast directly: 

```suggestion
                AgroalDataSource agroalDataSource = (AgroalDataSource) dataSource;
```

File: integration-tests/flyway/src/test/java/io/quarkus/it/flyway/FlywayFunctionalityTest.java
Comment (gastaldi) [inline integration-tests/flyway/src/test/java/io/quarkus/it/flyway/FlywayFunctionalityTest.java:59]:
```suggestion
    @DisplayName("Returns if the created-by user is different from quarkus")
```

**Review by gastaldi — APPROVED** [review]
LGTM, but I'll let @gsmet have the final say on this

**Comment by gastaldi** [comment]
@gsmet kind reminder to have a look when possible ;)

**Review by geoand — APPROVED** [review]

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java
Comment (gsmet) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:50]: I saw in several pull requests of yours that you add a new line after the javadoc. Let's not do that for the next ones. Thanks!

**Review by gsmet — CHANGES_REQUESTED** [review]
I spotted a few minor issues, I will fix them.

**Review by gsmet — APPROVED** [review]
I applied a few minor changes. Things should be good to go now. Thanks!

**Comment by chrisruffalo** [comment]
@gsmet @gastaldi is there any chance for this to be backported into 2.15 or 2.16?

**Comment by gastaldi** [comment]
@chrisruffalo I marked it to be backported to 2.16.2.Final. 2.15 is no longer updated

**Comment by gastaldi** [comment]
There are no plans for a 2.7, the next release is 3.0. We're working hard on it

**Comment by chrisruffalo** [comment]
@gastaldi that makes sense, thanks.

**Pull request merged**
